### ISSUE TITLE
D7 to D10 Article Migrations for CCOM

### ISSUE THREAD

**Issue by ethan-fry** [uiowa/uiowa#6640]
Looking through the source code for site now I have found the engineering_migrate module, we are hoping to provide our group of subsites currently on D7 CCOM the ability to migrate their current content over from D9 SiteNow.  Would this module be a good starting point to begin looking into implementing a migration feed? We would be looking to only migrate over article and basic page content types with the migration.

**Update - User Story**
https://github.com/uiowa/uiowa/issues/6640#issuecomment-1775593940

**Comment by cory-skeers** [comment]
Re-purposing a previous writeup. Some of it may have since changed, but it'll be a closer match to how migrations work currently and might give a better starting point, as the engineering migrate was an early migration.

The bulk of the code can be found at:

https://github.com/uiowa/uiowa/tree/main/docroot/modules/custom/sitenow_migrate 

The majority of our current migrations are custom in some manner, but their modules always include this as their basis.

 

There’s a basic form to fill out database connection information:

https://github.com/uiowa/uiowa/blob/main/docroot/modules/custom/sitenow_migrate/src/Form/MigrateSettingsForm.php 

We use this as every migration has a different source, and it may or may not be a local database connection or connection to the Acquia Cloud application. Not as necessary if you’re not running many different migrations.

 

Sitenow Migrate has a series of Source classes:

https://github.com/uiowa/uiowa/tree/main/docroot/modules/custom/sitenow_migrate/src/Plugin/migrate/source 

BaseNodeSource.php, in particular, is the class we extend for the majority of our migrations. It extends Core’s d7 Node Class to add some additional functionality.

Couple common features we’ve added here:

Convert previous unpublished/published boolean to a moderation state value
Helper function to fetch all values of a multi-value field
Helper function to fetch any associated URL aliases for the migrated node
Definitions for pre- and post-import processes that utilize Core’s ImportAwareInterface, and allow you to call functions prior to and following a migration
Helper function to clear caches and resetting storage--helps with some of the PHP memory limit errors that arise for larger migrations. There's an unknown issue in our file download function that isn't releasing memory properly
Helper functions to fetch a summary from a field, or create one if one doesn’t already exist
Helper function to grab the id of the last node migrated (assuming they were pulled in order)
 

AliasRedirect.php

A simple Source class that can be used with a YML definition file to create redirects from old aliases. Pretty much all of our redirect migrations can use this directly with an almost copy-paste YML file.

 

ProcessMediaTrait

A trait with some common media processing functions that can be added onto a migration. Some of its functionality is specific to our system, like specifying image view modes, but some of it could be used for other file/media setups.

It has some helper functions to fetch file directory paths, construct new ones, etc.

Three methods in particular get used a lot: replaceInlineFiles, processImageField, and replaceInlineImages.

processImageField takes an image field file id from the d7 site, determines what image it’s supposed to use, and either

Finds the file in the d8 database if it’s been moved over already, and returns the new media id to use in the field or
Downloads the file, creates a new media entity, and returns the media id to use in the field
replaceInlineFiles uses regular expressions to scan d7 text fields for entity embeds, follows a similar process to processImageField to find/create a media entity, and then manually constructs and replaces the old d7 entity embed code with new d8 media embed code.

 

replaceInlineImages is similar to replaceInlineFiles, but takes a different approach, and because of it only works for images. It takes the migrated text and runs it through a renderer to create rendered HTML. It then parses the newly created DOM to look for any image references, updates them to the new file locations (downloading images and creating new media entities if necessary), and then re-serializes the DOM back to a text string which can then be attached to the appropriate d8 field.

 

All these processes use the downloadFile and createMediaEntity functions extensively: https://github.com/uiowa/uiowa/blob/0ebb4333597f56fc8a5b68ca93f9d3121d1d7a0d/docroot/modules/custom/sitenow_migrate/src/Plugin/migrate/source/ProcessMediaTrait.php#L343-L499

The downloadFile method and the helper functions used around calls to it check if a file exists in the d8 file directory and if not downloads it directly from the source site. File URIs are constructed manually using the source URL as well as the d7 file database tables. We only pull public files, then, as the migration needs to be able to directly access the file URI for download.

createMediaEntity does just that--once a file is in the file directory, to actually use it in d8 there needs to be a media entity. This method creates that entity and returns the newly created media entitiy's id.

 

linkReplaceTrait has a few helper functions.

For clients who have a direct mapping of content that’s been migrated previously, we use manualLookup and preLinkReplace to scan content for relative links and, if it points to content that’s already been moved over, update them so they point to the new location. It parses the DOM like replaceInlineImages.

Otherwise there’s postLinkReplace, which queries the database using a regular expression to look for pages which have suspected broken links (node/ format links) after migration. This can use the manualLookup as well, or manualLookup can be replaced to look at the migrate_map_ tables and attempt to fix links that are pointing to content which has just been migrated. This method uses the database query to get a list of potential pages that need updated links, but then parses the DOM as the actual find/replace methodology in order to avoid needing to perform database INSERTs on already migrated content.

Lastly, reportPossibleLinkBreaks only scans the database for any links in /node format and outputs the list of nodes following migration. This is largely used in our article migrations to offer reports of links in articles that point to non-article content that doesn't have a manual map constructed, or that points to content that has not yet been moved, and therefore doesn't have a new destination.

 

Articles.php, Pages.php, and People.php are all specific Source classes to migrate content from standard SiteNow v1 into SiteNow v3.

Files.php is specific as well, but is more general. We use it occasionally, but it's an all-or-nothing approach to files. More commonly, we migrate files on-demand as part of another migration.

 

We used to use postRowSaveEvent event subscriber more.

https://github.com/uiowa/uiowa/blob/dafd48b8e42274743ae2bd29a793ef464fa1f6a0/docroot/modules/custom/sitenow_migrate/src/EventSubscriber/PostRowSaveEvent.php (previous revision)

It’s still used for the file migration. In the migration, files are brought over from d7 to d8, and the event listener then creates a media entity for the file after it has been saved.

Previously, we used it for dropping content into paragraphs as part of the SiteNow v2 migrations. The destination node doesn't get created until the very end of each row migration, so some processes like creating section and content paragraphs attached to the node can’t yet happen. This listener waited for the node to get created and saved as the row migration completes, then went back to grab the default-created section, create a new paragraph using the migrated content, and attach it to that newly created section. 

If content is going to be migrated into paragraphs and not direct node fields, an event listener like this will likely need to be created and expanded upon.

 

The Obermann migration is a pretty recently completed migration, and an example of how the base clases get extended (we have more examples available if it'd be helpful):

https://github.com/uiowa/uiowa/tree/main/docroot/sites/obermann.uiowa.edu/modules/obermann_migrate

The config/split directory contains the migration YML definitions: https://github.com/uiowa/uiowa/tree/main/docroot/sites/obermann.uiowa.edu/modules/obermann_migrate/config/split

It’s separated into two—the article migration itself, and then a followup migration to create redirects out of the old aliases. The YML migration definitions are fairly straightforward destination:source mappings, but you can fit some pretty complex stuff into them using plugins.

The Articles.php (https://github.com/uiowa/uiowa/blob/main/docroot/sites/obermann.uiowa.edu/modules/obermann_migrate/src/Plugin/migrate/source/Articles.php) extends our BaseNodeSource and uses both the ProcessMediaTrait and LinkReplaceTrait. It had a few extras, like converting old node entity references to programs into taxonomy term references. In this case, taxonomy terms had already been created, but in other migrations we’ve manually created new taxonomy terms here.

 

Sitenow P2LB

It’s not directly a migration, but the Sitenow P2LB module might be useful for a reference: https://github.com/uiowa/uiowa/tree/p2lb/docroot/modules/custom/sitenow_p2lb 

It is a module designed to convert SiteNow v2 sites (which use Paragraphs) into SiteNow v3 (which use Layout Builder), and so has some processes for handling paragraphs attached to nodes.

This is the main driver for an individual node: https://github.com/uiowa/uiowa/blob/24004c7c3f20b671fff09f9a1eb569642bcd6a0f/docroot/modules/custom/sitenow_p2lb/sitenow_p2lb.module#L31-L85

It calls a short helper function that can be used to either grab sections attached to a node or paragraphs attached to sections:

https://github.com/uiowa/uiowa/blob/24004c7c3f20b671fff09f9a1eb569642bcd6a0f/docroot/modules/custom/sitenow_p2lb/sitenow_p2lb.module#L87-L115

There is some manual section processing to determine the “shape” of the content. Part of this is specific to getting it ready to be moved to Layout Builder, but the initial processing could be used as a reference if you’re going from Paragraphs to Paragraphs, or from node content to Paragraphs. In some cases it may be doing things "backwards," as we're fetching info from Paragraphs rather than setting it.

https://github.com/uiowa/uiowa/blob/24004c7c3f20b671fff09f9a1eb569642bcd6a0f/docroot/modules/custom/sitenow_p2lb/sitenow_p2lb.module#L117-L351

**Comment by ethan-fry** [comment]
@cory-skeers 

With the work you are currently doing for the media migrations it came up in a discussion with @briand44 that he was unsure about what would happen with the article migrations if the media already exists from the media work you are currently doing.

@maxwell-freund

**Comment by cory-skeers** [comment]
> @cory-skeers
> 
> With the work you are currently doing for the media migrations it came up in a discussion with @briand44 that he was unsure about what would happen with the article migrations if the media already exists from the media work you are currently doing.
> 
> @maxwell-freund

The way the article migrations work, every time a file is encountered it checks if the file (and media entity) exist on the destination site. If they do, that media entity gets referenced, and if not, then the file gets downloaded, a media entity gets created for it, and then that newly created media entity is used.

For the articles migration, this is specifically to avoid duplicating files that are used on multiple articles, but has the side benefit of allowing it to play nicely with previous files migrations. and it'll search for previously migrated files to reference and only pull in new ones if they weren't previously migrated.

**Comment by maxwell-freund** [comment]
@cory-skeers thanks for the update and details.

Great to hear the article migration takes into account previously migrated files, so both could run on a given site without a negative impact.

@ethan-fry @cory-skeers are there outstanding but known steps to validating this article migration can properly pull content off CCOM articles that need to be groomed/prioritized in coming sprints?

Anything our UI Health Care dev team needs to bring forward, or make accessible off CCOM D7 to be able to do that?

**Comment by briand44** [comment]
# Story

As a UIHC MarCom developer, I would like to be able to run a D7->v3 article migration for some CCOM sites without having to create a site split for each site we migrate articles for.

# Additional Info

Max is going to provide a field mapping from CCOM articles to v3

Likely need to migrate from local to local and sync similar to file migrations.

We'll want to document process if not already captured somewhere.

Should this be a separate feature split so it can be turned off once the migration is complete?

# Proposed Solution

- Create a default CCOM article migration
- Map fields
- If D7 External URL or Iowa Now URL filled out -> Fill URL Value into Source Link, Set "Link Directly To Source" = True.
- Just set site-wide default for ALL article images, accepting that some might not be the best fit for the widescreen, desktop displays
- migrate tags
- create redirects
- document process to run migration

**Comment by maxwell-freund** [comment]
@briand44 and others. Attached is an excel that maps out the D7 Article fields for CCOM and my best attempt to align with the field as I see it in the D9 admin for SiteNow.

If there was anything that was concerning or I had questions on, I added a column for those thoughts too.
[article migration mapping.xlsx](https://github.com/uiowa/uiowa/files/13077685/article.migration.mapping.xlsx)

**Comment by cory-skeers** [comment]
<details>
<summary>
Mapping
</summary>

D7 Label | D7 Machine Name | D7 Field Type | D9 Field (Admin View) | Thoughts / Questions
-- | -- | -- | -- | --
Body | body | Long text and summary | Body | Confirming media entity   references that are embedded in body will come through?
Date | field_date | Date | Authored On |  
External URL | field_article_external_url | Link | Source Link | Assuming Source Link is only   used for offsite links? If so, this is the right mapping
Gallery | field_article_gallery | Image | Gallery |  
Image | field_image | Image | Featured Image |  
IowaNow URL | field_article_iowanow_url | Link | Source Link | Assuming Source Link is only   used for offsite links? If so, this is the right mapping
Meta tags | metatags | Meta tag module form elements. | Basic Tags | Not sure this maps, but closest   matching field so placed here.
Publication Source | field_article_publication | Text | Organization | Publication Source was only used   to acknowledge outside orgs. Organization reads similarly in D9, but would   want confirmation before mapping
Tags | field_tags | Term reference | Relationships > Tags | Most sites did not leverage   tagging taxonomies, but where they exist, would be nice to map through to D9   tags.
Title | title | Node module element | Title |  
Type | field_article_type | List (text) | Not Needed | Likely not needed. This field   managed showing/hiding redirect URLs on the D7 admin. Is extraneous in D9, as   those fields just exist on all articles if needed.
URL path settings | path | Path module form elements | URL Alias |  
URL redirects | redirect | Redirect module form elements | …Not Possible? | How do existing redirects that   are pointing to a D7 node ID work when they are moving over to a new site   with a new node ID value for this article? Assuming those old redirects   cannot come along for the ride with this migration, but would like confirmation   if it is possible or not.

</details>

Couple notes:

> Body: Confirming media entity   references that are embedded in body will come through?

Yes--If a files migration is run prior to the article migration, the migration will look up the file, find the media entity on the D9/10 site, and wire it up. If a files migration hasn't been run prior, it's a similar process, but the migration will also be expected to download each file and create a media entity for it in d9/10 if it hasn't already done so (ie if another article used the same image, it will only download if it hasn't already for a previously migrated article)

> External URL / IowaNow URL: Assuming Source Link is only   used for offsite links? If so, this is the right mapping

Yes. Additionally, there is a "link directly to source" option that we can set to True or False with this. 

If True, then anywhere the article appears as a teaser (such as a general /news page), it will link to that external source rather than the article on the site. 

If False, the external source will display, but the link will go to an article on the site. A lot of times this contains some summary of the article (prof won this award, featured in the DI) and the link is intended as sort of a "read more" option.

Either case has its use, and we can default it to either for all articles that have an external url or iowa now url.

> Featured image

In addition to mapping here, there is the option of whether to display the image on the page (if not, it will only be used in article teasers) and, if so, whether to use the large, medium, small, or site default for its size.

Three ways we have approached this have been to

1. Just set site-wide default for ALL article images, accepting that some might not be the best fit for the widescreen, desktop displays
2. Hide ALL article images from migration, assuming that few or none of the source images are a good fit for the widescreen, desktop displays
3. Have a quick image size check, and if they are above a certain threshold for pixel size and/or in a certain aspect ratio (3:2 or 'wider') then display, else hide.

> Publication Source: Publication Source was only used to acknowledge outside orgs. Organization reads similarly in D9, but would want confirmation before mapping

Yes--organization is a bit of a catchall, but that's the general intention.

> URL Redirects: How do existing redirects that are pointing to a D7 node ID work when they are moving over to a new site with a new node ID value for this article? Assuming those old redirects cannot come along for the ride with this migration, but would like confirmation if it is possible or not.

We can bring over redirects.

Most migrations we do come in two parts, the article migration and a "redirect" migration. This second isn't a typical redirect migration, but rather sets up redirects on the new site matching the old site's URL pattern.

So, `node/41` on the original site has the alias `news/article-title`, but the automatic alias created on the new site (which is now housed at `node/72`) is `news/2023/10/article-title`. So this second migration then also creates a redirect for `news/article-title` to `node/72`.

We can set up a third migration (or as part of the first migration) that would also check for other redirects and create them on the new site, as the migration keeps track of what node from the old site corresponds to what that node is on the new site.

The main reason we've avoided doing this in some cases in the past is we've found a lot of the redirects to be old and never getting used, and went instead the route of: match the old article alias and create a redirect and then monitor GA after the new site is launched and add redirects as needed for any 404 occurrences.

**Comment by maxwell-freund** [comment]
Thanks @cory-skeers for the thorough review. Read through your notes, the only ones I have thoughts to expand upon are as follows:

_External URL / IowaNow URL: Assuming Source Link is only used for offsite links? If so, this is the right mapping

Yes. Additionally, there is a "link directly to source" option that we can set to True or False with this.

If True, then anywhere the article appears as a teaser (such as a general /news page), it will link to that external source rather than the article on the site.

If False, the external source will display, but the link will go to an article on the site. A lot of times this contains some summary of the article (prof won this award, featured in the DI) and the link is intended as sort of a "read more" option.

Either case has its use, and we can default it to either for all articles that have an external url or iowa now url._

Knowing how the External URL and IowaNow URL function on Drupal7 CCOM, it is intended to function in the way you describe for the "True" scenario. Simply appearing as a teaser on a news search results page, etc. and when clicked, navigating directly to the article on the other website.

Therefore, for this migration, I think we can assume the logic:

If D7 External URL or Iowa Now URL filled out -> Fill URL Value into Source Link, Set "Link Directly To Source" = True.

---

Regarding Image -> Feature Image.

Your first option:

_Just set site-wide default for ALL article images, accepting that some might not be the best fit for the widescreen, desktop displays_

Certainly feels acceptable, since the expectation is at bare minimum, the author checks each article once migrated for basic aesthetics, etc. And if they come across an article where an image is poorly optimized in the this way, they can address it in that article in D9/10.

----

Regarding redirects. Your explanation makes sense. I think the 2nd migration you outline here:

_So, node/41 on the original site has the alias news/article-title, but the automatic alias created on the new site (which is now housed at node/72) is news/2023/10/article-title. So this second migration then also creates a redirect for news/article-title to node/72._

That cleans up the old alias URL to the new alias, would be great to have.

Your further explanation on avoiding the 3rd migration to check for other redirects and create them on the site, the level of effort associated with it with likely, fairly minimal impact, is totally fair. And monitoring 404s for that post launch feels like a completely acceptable alternative to a heavy dev lift on the front end.

**Comment by cory-skeers** [comment]
Looking at medicine.uiowa.edu/internalmedicine, there are some articles like https://medicine.uiowa.edu/internalmedicine/node/8176, which feature a date in the body field (Thursday, May 21, 2015) which doesn't match the article's creation date (Friday, June 24, 2016) and there is no date listed in the Date field.

This results in two (separate) dates being displayed:
<img width="1358" alt="image" src="https://github.com/uiowa/uiowa/assets/49074566/4e9dfad9-8630-4656-ba02-d33fb0f54973">

I'm thinking we can do a little bit of cleanup stripping the date string out of the body (if one is found alone at the start of the content), and also use it as a backup created date if the date field doesn't have a value.

But going to check first if this is a common occurrence, or only happens to be the articles I've been using with internal med in testing.

**Comment by maxwell-freund** [comment]
@cory-skeers thanks for bringing this up and asking the question.

Looking at the date back in 2015, I think this might have been a past practice that we no longer do. I have no real good justification for why a date would have been put in the body, instead of leveraging the date field for this piece.

Our standard approach to articles on CCOM is to simply let the publish date stored in the system date field, to be shown on the page as seen here: https://medicine.uiowa.edu/content/athletic-training-residency-program-granted-initial-accreditation

Now if you see this practice having happened many times in past years, I welcome thoughts on how to handle this variation...

But I certainly would not expect the export/import to be scraping the body of the article, find a date in body copy, and adjust accordingly.

That feels like manual cleanup that could be put on an author once the article is migrated into D9...

**Comment by cory-skeers** [comment]
> Looking at the date back in 2015, I think this might have been a past practice that we no longer do. I have no real good justification for why a date would have been put in the body, instead of leveraging the date field for this piece.
> 
> Our standard approach to articles on CCOM is to simply let the publish date stored in the system date field, to be shown on the page as seen here: https://medicine.uiowa.edu/content/athletic-training-residency-program-granted-initial-accreditation
> 
> Now if you see this practice having happened many times in past years, I welcome thoughts on how to handle this variation...
> 
> But I certainly would not expect the export/import to be scraping the body of the article, find a date in body copy, and adjust accordingly.
> 
> That feels like manual cleanup that could be put on an author once the article is migrated into D9...

From what I've been seeing, it looks like articles from before ~November 2020 usually have it. For a site like medicine.uiowa.edu/pathology, that's 230 of the 300 articles. There are occasional ones that have cropped up later, like https://medicine.uiowa.edu/pathology/node/5046 (feb 3, 2022). 

My thought was to add something like https://regex101.com/r/eK8xyS/1 as a regex to strip it out. It wouldn't be a 100% solution if formats varied, but it would cover the examples I've come across so far without risking legitimate dates within the body.

**Comment by maxwell-freund** [comment]
@cory-skeers this one is a great example of duplication: https://medicine.uiowa.edu/pathology/node/5046 and 100% agree with stripping it out of the body. It is extraneous there.

So just to confirm the possible scenarios and ways to address:

1. If Date field is filled out and the REGEX check finds a date also in the body -> Attempt to strip it out. (pathology article example)
2. If Date field is filled out and the REGEX does not find a date to strip out -> no issue, proceed as planned
3. If Date Field is NOT filled out and there is a date field found in the body (the 2015 internal medicine example) -> What do we do here? Still strip it out? Strip it out and populate the Date field...? Help me understand this one further.
4. If a Date Field is not filled out and the REGEX does not find a date in the body -> There really is nothing to do here. There's no date to migrate... It migrates over using the node's authored on date? 
5.

**Comment by cory-skeers** [comment]
@maxwell-freund I think the main thing is I'd like to strip the date out of the body in these cases, since there will always be a date shown automatically. Since the form is consistent as far as I've seen, it's a pretty low-impact addition to take care of that.

We can either leave it at that, or try and use the date as well.

> So just to confirm the possible scenarios and ways to address:
> 
> 1. If Date field is filled out and the REGEX check finds a date also in the body -> Attempt to strip it out. (pathology article example)
> 2. If Date field is filled out and the REGEX does not find a date to strip out -> no issue, proceed as planned
> 3. If Date Field is NOT filled out and there is a date field found in the body (the 2015 internal medicine example) -> What do we do here? Still strip it out? Strip it out and populate the Date field...? Help me understand this one further.
> 4. If a Date Field is not filled out and the REGEX does not find a date in the body -> There really is nothing to do here. There's no date to migrate... It migrates over using the node's authored on date?

Just stripping the date if found covers scenarios 1, 2, and 4. And in the fourth case, it will just migrate over with the authored on date.

For scenario 3, my thinking is that if there was a date written in the body, it's likely more accurate than the authored on date since someone intentionally put it there, and so we would

- strip it out of the body
- use it to set the date field rather than using the authored on date

Mostly I'd like to make sure we strip it out to avoid the erroneous date, and it seems like an easy win to save some manual cleanup, but if there's also a rule we can use to go ahead and use it as a published date in some cases, it might be able to do double duty.

Another way to look at it would be that we'd strip out the date in all cases if it's there, and then:

1. If there's a Date field value, use that
2. If there's no Date field value, but there's a date listed at the beginning of the body, use that
3. If there are neither, use the authored on date

**Comment by maxwell-freund** [comment]
@cory-skeers thanks for walking through these scenarios.

I think your summation at the end of the 3 scenarios following stripping out an in-body mention all make sense and I support.

I do think the scenario 2 you list: If there's no Date field value, but there's a date listed at the beginning of the body, use that

is an assumption I think makes sense, as you say, someone put it there intentionally, so it likely is the more accurate pub date than a possible authored on system variable.

That said, this feels like a 100% nice to have. Is this going to be a significant lift to use the in-text value in this way? If so, I am perfectly comfortable falling back to the authored on date instead...

Thanks.

**Comment by cory-skeers** [comment]
I'm taking a look at the `metatags`, and it looks like it'll be a little complicated to get set up, and that it's not used much.

There are meta values for articles, but they're mostly what get constructed by the system. I checked the databases for medicine.uiowa.edu/internalmedicine, medicine.uiowa.edu/pathology, and medicine.uiowa.edu/anesthesia, and the only article metatag I've found between the three is for https://medicine.uiowa.edu/pathology/node/4261, which has a canonical url set for https://medicine.uiowa.edu/content/ui-pathology-professor%E2%80%99s-nobel-connection.

Since something like that could definitely be useful, but is probably a good candidate for manual cleanup/addition, I'm going to look at logging it as a message during migration rather than mapping directly.
